larq_compute_engine/core/bitpack.h

```diff
@@ -184,18 +180,60 @@ template <class TIn>
 inline void bitpack_matrix(const TIn* input, const std::size_t input_num_rows,
                            const std::size_t input_num_cols, TBitpacked* output,
                            const std::int32_t zero_point = 0) {
+  // Calculate the size of the bitpacked rows
+  const std::size_t output_num_cols = GetBitpackedSize(input_num_cols);
+
+  // The public API has an Int32 zero-point, to match the TFLite spec, but
+  // bitpacking is trivial if the zero-point is outside the TIn representable
+  // range.
+
+  if (zero_point <= std::numeric_limits<TIn>::lowest()) {
+    // All values must represent >= 0, so the bitpacked bits will all be 0.
+    std::fill(output, output + input_num_rows * output_num_cols, TBitpacked(0));
+    return;
+  }
+
+  if (zero_point > std::numeric_limits<TIn>::max()) {
+    // All values must represent < 0, so the bitpacked bits will all be 1.
+    //     Unlike the >= 0 case above, this needs more caution, because while
+    // the bitpacking bits will all be 1, any padding bits, if applicable, still
+    // must be zero.
+    if (input_num_cols % bitpacking_bitwidth == 0) {
+      // First, the easy case where there's no padding: just binary ones.
+      std::fill(output, output + input_num_rows * output_num_cols,
+                ~TBitpacked(0));
+    } else {
+      // With padding, all elements of each output column will be binary ones,
+      // except the final element, which will have a prefix of ones with a
+      // suffix of zeroes. We pre-compute it outside the loop.
+      TBitpacked value_with_padding = 0;
+      for (int i = 0; i < input_num_cols % bitpacking_bitwidth; i++) {
+        value_with_padding |= TBitpacked(1) << i;
+      }
```

Comment:
I think you can create a `00..000111..11` bitstring by either doing `(~0u) >> n` (with `n` the number of 0s, and `u` to make sure it's an unsigned shift that actually adds 0s instead of 1s) or by doing `(1 << n) -1` with `n` the number of 1s:
```suggestion
      TBitpacked value_with_padding = (1 << (input_num_cols % bitpacking_bitwidth)) - 1;
```

Reply:
That's a nice suggestion, thanks! Will do this in #494.